zip-0228.rst

```diff
@@ -2,12 +2,530 @@
+This Action, paying fees, in ZEC, to the matcher, is generated using the matcher's details to generate output notes, since we assume, generally, that the matcher is known by all trading parties.
+In the P2P case the matcher is simply the second party out of the two (See the `Fees`_ section for more details).
+
+ZIP 226 [#zip-0226]_, that this ZIP builds on, introduces changes to the NU5 circuit to include support for different Assets. 
+The structure there requires that the input and output notes of the OrchardZSA Action must have the same asset base.
+This ZIP continues with the same idea, i.e. the manipulation of a single Asset Base per Action.
+Therefore, Swap Orders are expected to generally have at least three Actions:
+
+- One Action for the Proposed Asset.
+- One Action for the Desired Asset.
+- One Action for the fees (:math:`\mathsf{ZEC}`).
+
+Security and Privacy Considerations
+===================================
+
+Protection against Replay attacks
+---------------------------------
+
+We consider whether our change from signing the SIGHASH in the spend authorization signature to signing the Action Group Hash opens any possibilities of replay attacks.
+
+This is prevented by the use of the updated SIGHASH in the binding signature. 
+If an adversary tries to extract an Action Group and associated Spend Authorization Signature from a transaction on the network to replay it within another transaction - which would potentially be detrimental to the matcher as it could make their bundle invalid (if the replayed Action Group gets included first on-chain) - the adversary will also need to be able to generate a binding signature on their replayed transaction, which is not possible without knowing the :math:`\mathsf{bsk}` associated with the Action Group being replayed.
+The :math:`\mathsf{bsk}` is sent within the Swap Order, which is assumed to be communicated over a secure channel off-chain between the parties. 
+This precludes the possibility of replay attacks. 
+
+Non-Malleability of the Time Limit
+----------------------------------
+
+We protect against the malleation of the time limit field by a malicious matching party (in order to take advantage of the "free option" that it could obtain by extending the validity of an Order) by including the time limit inside the Action Group Hash that is signed using the Spend Authorization Signature (see more details in `Rationale for Time Limit`_).
+The security of the Spend Authorization Signature and the collision resistance of the BLAKE2b-256 hash then ensures that the time limit remains the same as the one mandated by the creator of the Swap Order.
+
+Other Considerations
+====================
+
+Fees
+----
+
+The primary goal of a fee mechanism is twofold: 
+
+- It should disincentivize malicious users from flooding the network with spam Swap offers that are unlikely to be matched.
```

Comment:
Re: this.

I don't think the fee mechanism **directly** disincentizes network flooding properly speaking - I think it depends on the overall setting.

Any trader could send/flood the Matcher (or Matcher network) with lots of rogue Swap Orders that are invalid. The fees don't prevent this and don't incur any **direct** monetary cost on the sender of the Swap Orders (**indirect** costs are incurred by the sender in the form of the cost of the computational resources needed to flood the network, not because of the fees which are only paid when the Swap Order is matched and the bundle is settled on chain).

Usually, the Matcher will (depending on the setting), put some mechanisms in place to limit flooding, e.g.
- Relay rules (if applicable) where peer nodes in the p2p network abud by specific rules for relaying transaction. E.g. if transactions have too low of a fee they may not be relayed (i.e. here 'fees' could play a role in limiting flooding)
- Ban policies/IP address blocking (https://en.wikipedia.org/wiki/IP_address_blocking) where peers that send many invalid transactions or spam the network get banned (by other peers), limiting their ability to broadcast/send requests

Etc.

Reply:
I guess it all depends which 'network' we talk about here (the Zcash network or the Relay/Matcher network?). Given that the sentence goes on to say "flooding the network with spam Swap offers that are unlikely to be matched." -> it feels like we're talking about flooding the Matcher (because usually only Bundles will make it to the Zcash network, not Orders which are sent off-chain)

Reply:
Updated Swap offer to Swap Order in #62 , that was an error, my bad.

Reply:
I will reword the Fees section based on Pablo's and Antoine's comments in #52, and account for these comments.